src/test/scala/chiselTests/InstanceSpec.scala

```diff
@@ -0,0 +1,444 @@
+          w := in + 1.U
+          out := w
+        case None =>
+          out := in + 1.U
+      }
+    }
+    @instance
+    class AddTwo extends MultiIOModule {
+      @public val in  = IO(Input(UInt(32.W)))
+      @public val out = IO(Output(UInt(32.W)))
+      val template = Template(new AddOne(true))
+      @public val i0 = Instance(template)
+      @public val i1 = Module(new AddOne(true))
+      //@public val x = 10 //ERRORS!!!
+      i0.in := in
+      i1.in := i0.out
+      out := i1.out
+    }
+  }
+  object Annotations {
+    case class MarkAnnotation(target: ReferenceTarget, tag: String) extends SingleTargetAnnotation[ReferenceTarget] {
+      def duplicate(n: ReferenceTarget): Annotation = this.copy(target = n)
+    }
+    case class MarkChiselAnnotation(d: Data, tag: String) extends chisel3.experimental.ChiselAnnotation {
+      def toFirrtl = MarkAnnotation(d.toTarget, tag)
+    }
+    def mark(d: Data, tag: String): Unit = annotate(MarkChiselAnnotation(d, tag))
+  }
+//  object HierarchyExamples {
+//    //implicit def convert(m: AddOne): Instance[AddZeroInterface] = { }
+//    @instance
+//    final class AddOne extends MultiIOModule with AddInterface {
+//      @public val in  = in
+//      @public val out = out
+//      @public val innerWire = Wire(UInt(32.W))
+//      innerWire := in + 1.U
+//      out := innerWire
+//    }
+//    @instance
+//    class AddTwo extends MultiIOModule with AddInterface {
+//      @public val in  = IO(Input(UInt(32.W)))
+//      @public val out = IO(Output(UInt(32.W)))
+//      @public val innerWire = Wire(UInt(32.W))
+//      @public val innerModule = Module(new Interfaces.AddZero)
+//      innerWire := in + 2.U
+//      out := innerWire
+//    }
+//  }
+}
+
+
+
+class InstanceSpec extends ChiselFlatSpec with Utils {
+  "Template/Instance" should "enable instantiating the same instance multiple times" in {
+    import InstanceSpec.Examples._
+    import InstanceSpec.Annotations._
+    class AddOneTester extends BasicTester {
+      val template: Template[AddOne] = Template(new AddOne(true))
+      val i0: Instance[AddOne] = Instance(template)
+      val i1: Instance[AddOne] = Instance(template)
+      i0.in := 42.U
+      i1.in := i0.out
+      i0.innerWire.map(x => mark(x, "Jack Was Here"))
+      chisel3.assert(i1.out === 44.U)
+      stop()
+    }
+    val (output, annotations) = (new ChiselStage).emitChirrtlWithAnnotations(gen = new AddOneTester, args = Array("--full-stacktrace"))
+    annotations.toSeq should contain (MarkAnnotation(Target.deserialize("~AddOneTester|AddOneTester/i0:AddOne>innerWire").asInstanceOf[ReferenceTarget], "Jack Was Here"))
+    assertTesterPasses(new AddOneTester)
+  }
+  "Template/Instance" should "enable instantiating nestingly, with modules" in {
+    import InstanceSpec.Examples._
+    import InstanceSpec.Annotations._
+    class AddOneTester extends BasicTester {
+      val template = Template(new AddTwo)
+      val i0 = Instance(template)
```

Comment:
Have you considered overloading the `Module.apply` method instead of creating a new `Instance.apply`? That way users have to remember fewer keywords.

Reply:
Hmm I didn't, but that's a good idea to consider.

Reply:
I don't think you can overload here, `Module.apply` is of type `T <: BaseModule => T`, `Instance.apply` is of type `T <: BaseModule => Template[T]`. You can only overload when the argument type is different and it's the same for both of these.

Reply:
I wasn't thinking of `Module(new AddTwo)` but instead of `Module(Template(new AddTwo))`

Reply:
Got it, yeah that should be doable.